pkg/controller/gc_controller.go

```diff
@@ -46,100 +41,63 @@ const (
 	gcFailureBSLReadOnly     = "BSLReadOnly"
 )
 
-// gcController creates DeleteBackupRequests for expired backups.
-type gcController struct {
-	*genericController
-
-	backupLister              velerov1listers.BackupLister
-	deleteBackupRequestLister velerov1listers.DeleteBackupRequestLister
-	deleteBackupRequestClient velerov1client.DeleteBackupRequestsGetter
-	kbClient                  client.Client
-	frequency                 time.Duration
-
-	clock clock.Clock
+// gcReconciler creates DeleteBackupRequests for expired backups.
+type gcReconciler struct {
+	client.Client
+	logger logrus.FieldLogger
+	clock  clock.Clock
 }
 
-// NewGCController constructs a new gcController.
-func NewGCController(
+// NewGCReconciler constructs a new gcReconciler.
+func NewGCReconciler(
 	logger logrus.FieldLogger,
-	backupInformer velerov1informers.BackupInformer,
-	deleteBackupRequestLister velerov1listers.DeleteBackupRequestLister,
-	deleteBackupRequestClient velerov1client.DeleteBackupRequestsGetter,
-	kbClient client.Client,
-	frequency time.Duration,
-) Interface {
-	c := &gcController{
-		genericController:         newGenericController(GarbageCollection, logger),
-		clock:                     clock.RealClock{},
-		backupLister:              backupInformer.Lister(),
-		deleteBackupRequestLister: deleteBackupRequestLister,
-		deleteBackupRequestClient: deleteBackupRequestClient,
-		kbClient:                  kbClient,
+	client client.Client,
+) *gcReconciler {
+	return &gcReconciler{
+		Client: client,
+		logger: logger,
+		clock:  clock.RealClock{},
 	}
-
-	c.syncHandler = c.processQueueItem
-	c.resyncPeriod = frequency
-	if c.resyncPeriod <= 0 {
-		c.resyncPeriod = defaultGCFrequency
-	}
-	logger.Infof("Garbage collection frequency: %s", c.resyncPeriod.String())
-	c.resyncFunc = c.enqueueAllBackups
-
-	backupInformer.Informer().AddEventHandler(
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    c.enqueue,
-			UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
-		},
-	)
-
-	return c
 }
 
-// enqueueAllBackups lists all backups from cache and enqueues all of them so we can check each one
-// for expiration.
-func (c *gcController) enqueueAllBackups() {
-	c.logger.Debug("gcController.enqueueAllBackups")
-
-	backups, err := c.backupLister.List(labels.Everything())
-	if err != nil {
-		c.logger.WithError(errors.WithStack(err)).Error("error listing backups")
-		return
-	}
-
-	for _, backup := range backups {
-		c.enqueue(backup)
-	}
+func (c *gcReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	s := kube.NewPeriodicalEnqueueSource(c.logger, mgr.GetClient(), &velerov1api.BackupList{}, defaultGCFrequency)
```

Comment:
This needs modification, because GC controller only needs to be triggered one time per fequency.
If there is multiple backups in cluster, GC reconcile will be triggered multiple times.
Better to add a filter function to only enqueue one backup.